cppapi/server/eventsupplier.cpp

```diff
@@ -766,11 +770,13 @@ bool EventSupplier::detect_and_push_archive_event(DeviceImpl *device_impl,
 
         if (except != NULL)
         {
+            omni_mutex_lock oml(detect_mutex);
             attr.prev_archive_event.err = true;
             attr.prev_archive_event.except = *except;
         }
         else
         {
+            omni_mutex_lock oml(detect_mutex);
             if (attr_value.attr_val_5 != NULL)
             {
                 attr.prev_archive_event.value_4 = attr_value.attr_val_5->value;
```

Comment:
* probably should be applied also to prev_change_event

Race is between polling thread (line 782 in PR, holds attr_mutex and event_mutex) and readthread (line 1837 in PR, holds detect_mutex):
https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/eventsupplier.cpp#L1830

```
==================                                                                                                                                                                                                                                   (1/29 results) [1399/14345]
WARNING: ThreadSanitizer: data race (pid=3222)
  Read of size 8 at 0x7b040000a320 by thread T8 (mutexes: write M1006):
    #0 Tango::EventSupplier::detect_change(Tango::Attribute&, Tango::EventSupplier::SuppliedEventData&, bool, double&, double&, Tango::DevFailed*, bool&, Tango::DeviceImpl*) /home/michal/Documents/cppTango/cppapi/server/eventsupplier.cpp:1837 (libtango.so.9+0x8a4898)
    #1 Tango::Attribute::fire_archive_event(Tango::DevFailed*) /home/michal/Documents/cppTango/cppapi/server/attribute.cpp:4484 (libtango.so.9+0x7a0c8c)
    #2 Tango::DeviceImpl::push_archive_event(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, double*, timeval&, Tango::AttrQuality, long, long, bool) /home/michal/Documents/cppTango/cppapi/server/dev_event.cpp:1764 (libtango.so.9+0x7dafda)
    #3 TestPushCrash_ns::readthread::run(void*) ../src/readthread.cpp:148 (TestPushCrash+0x34480)
    #4 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:459 (libomnithread.so.4+0x3f34)
    #5 <null> <null> (libtsan.so.0+0x296ad)

  Previous write of size 8 at 0x7b040000a320 by thread T9 (mutexes: write M1262, write M1004):
    #0 operator new[](unsigned long) <null> (libtsan.so.0+0x73fda)
    #1 _CORBA_Sequence<double>::allocbuf(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:84 (libtango.so.9+0x528964)
    #2 _CORBA_Sequence<double>::copybuffer(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:205 (libtango.so.9+0x525767)
    #3 _CORBA_Sequence<double>::length(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:59 (libtango.so.9+0x52298a)
    #4 _CORBA_Sequence<double>::_CORBA_Sequence(_CORBA_Sequence<double> const&) /usr/include/omniORB4/seqTemplatedecls.h:147 (libtango.so.9+0x5ac977)
    #5 _CORBA_Unbounded_Sequence<double>::_CORBA_Unbounded_Sequence(_CORBA_Unbounded_Sequence<double> const&) /usr/include/omniORB4/seqTemplatedecls.h:242 (libtango.so.9+0x5a4867)
    #6 _CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8>::_CORBA_Unbounded_Sequence_w_FixSizeElement(_CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8> const&) /usr/include/omniORB4/seqTemplatedecls.h:503 (libtango.so.9+0x599ceb)
    #7 Tango::DevVarDoubleArray::DevVarDoubleArray(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:238 (libtango.so.9+0x590c43)
    #8 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3040 (libtango.so.9+0x592162)
    #9 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #10 Tango::EventSupplier::detect_and_push_archive_event(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::Attribute&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Tango::DevFailed*, timeval*, bool) /home/michal/D
ocuments/cppTango/cppapi/server/eventsupplier.cpp:782 (libtango.so.9+0x89f797)
    #11 Tango::EventSupplier::detect_and_push_events(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::DevFailed*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, timeval*) /home/michal/Documents/cppTango/cppapi/server/
eventsupplier.cpp:269 (libtango.so.9+0x89d4f7)
    #12 Tango::PollThread::poll_attr(Tango::WorkItem&) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:1968 (libtango.so.9+0x90bd84)
    #13 Tango::PollThread::one_more_poll() /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:799 (libtango.so.9+0x9061ee)
    #14 Tango::PollThread::run_undetached(void*) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:192 (libtango.so.9+0x903276)
    #15 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:462 (libomnithread.so.4+0x3f6a)
    #16 <null> <null> (libtsan.so.0+0x296ad)
```

Race is also between polling thread (line 782 in PR, holds attr_mutex and event_mutex) and readthread here (no mutex):
https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/attribute.cpp#L4490-L4492

```
==================
WARNING: ThreadSanitizer: data race (pid=3222)
  Read of size 1 at 0x7b0800013628 by thread T9 (mutexes: write M1262, write M1004):
    #0 _CORBA_Sequence<double>::~_CORBA_Sequence() /usr/include/omniORB4/seqTemplatedecls.h:127 (libtango.so.9+0x52818d)
    #1 _CORBA_Unbounded_Sequence<double>::~_CORBA_Unbounded_Sequence() /usr/include/omniORB4/seqTemplatedecls.h:253 (libtango.so.9+0x524c2b)
    #2 _CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8>::~_CORBA_Unbounded_Sequence_w_FixSizeElement() /usr/include/omniORB4/seqTemplatedecls.h:516 (libtango.so.9+0x52196b)
    #3 Tango::DevVarDoubleArray::~DevVarDoubleArray() /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:233 (libtango.so.9+0x51b625)
    #4 Tango::AttrValUnion::_release_member() /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3281 (libtango.so.9+0x592cf5)
    #5 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3036 (libtango.so.9+0x592100)
    #6 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #7 Tango::EventSupplier::detect_and_push_archive_event(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::Attribute&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Tango::DevFailed*, timeval*, bool) /home/michal/Do
cuments/cppTango/cppapi/server/eventsupplier.cpp:782 (libtango.so.9+0x89f797)
    #8 Tango::EventSupplier::detect_and_push_events(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::DevFailed*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, timeval*) /home/michal/Documents/cppTango/cppapi/server/e
ventsupplier.cpp:269 (libtango.so.9+0x89d4f7)
    #9 Tango::PollThread::poll_attr(Tango::WorkItem&) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:1968 (libtango.so.9+0x90bd84)
    #10 Tango::PollThread::one_more_poll() /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:799 (libtango.so.9+0x9061ee)
    #11 Tango::PollThread::run_undetached(void*) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:192 (libtango.so.9+0x903276)
    #12 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:462 (libomnithread.so.4+0x3f6a)
    #13 <null> <null> (libtsan.so.0+0x296ad)

  Previous write of size 8 at 0x7b0800013628 by thread T8:
    #0 operator new(unsigned long) <null> (libtsan.so.0+0x73f0a)
    #1 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3040 (libtango.so.9+0x592150)
    #2 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #3 Tango::Attribute::fire_archive_event(Tango::DevFailed*) /home/michal/Documents/cppTango/cppapi/server/attribute.cpp:4505 (libtango.so.9+0x7a0d4b)
    #4 Tango::DeviceImpl::push_archive_event(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, double*, timeval&, Tango::AttrQuality, long, long, bool) /home/michal/Documents/cppTango/cppapi/server/dev_event.cpp:1764 (libtango.so.9+0x7dafda)
    #5 TestPushCrash_ns::readthread::run(void*) ../src/readthread.cpp:148 (TestPushCrash+0x34480)
    #6 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:459 (libomnithread.so.4+0x3f34)
    #7 <null> <null> (libtsan.so.0+0x296ad)
```